### PR TITLE
authz-service/migrations: add index, and require it being updated in PRs

### PR DESCRIPTION
**Pull request by srenatus**
👉 [This is the new index README](https://github.com/chef/automate/blob/44701c690998664a56c187ea4e0dd2833d5b546f/components/authz-service/storage/postgres/migration/sql/README.md)
ℹ️ With this index, Github will let us notice *before merging* if there's been another merged PR -- that one would have updated the index, and they'd work on the same line. (🤞 I hope that's how it works 😄)

Failures to commit the uptodate index are now reported on the make target.

### CODE REVIEW

**Comment by srenatus** [comment]
This won't fly, `{version}_{title}.up.{extension}` is required: https://github.com/mattes/migrate/blob/master/MIGRATIONS.md#migration-filename-format

**Comment by phiggins** [comment]
Do the numbers need to be strictly sequential, or is it enough that they are deterministically sortable? We could solve this by using a convention like using the current timestamp as the number, and have a script/make task to generate them like:

```
Petes-MBP:automate-chef-io pete$ date -u '+%Y%m%d%H%M%S'
20190606190656
```

**Comment by srenatus** [comment]
@phiggins you're right,
> timestamps at an appropriate resolution

are possible according to the docs linked above. And it's a much better solution than the index. :tada: closing this.

**Comment by phiggins** [comment]
Hehe, I don't know about _better_ necessarily, but fewer moving parts. :)

**Comment by srenatus** [comment]
👉 #544

**Comment by sdelano** [comment]
I believe there are specific reasons that we haven't used timestamp-based numbering for migrations in the past. @danielsdeleo do you happen to remember? I seem to recall there was a good explanation on the philosophy behind it somewhere on the Internet at some point in time :)

File: components/authz-service/Makefile [hunk 1]
```diff
@@ -14,10 +14,15 @@ ifdef CI
...
 
 all: lint build test
-
-static: lint
+static: lint ${MIGRATION_READMES}
 unit: build test
 
+.PHONY: ${MIGRATION_READMES}
+${MIGRATION_READMES}:
+	echo '## SQL migrations' > $@
```
[the remark this answers is not in the document]
Comment (srenatus) [reply components/authz-service/Makefile]: So, it wants a newline after the `##` line? OK.

File: components/authz-service/Makefile [hunk 2]
```diff
@@ -14,10 +14,15 @@ ifdef CI
...
+	for file in $$(dirname $@)/*.up.sql; do f=$$(basename $$file); echo "- [\`$$f\`]($$f): $$(sed '1s/--\/ //;q' "$$file")" >> $@; done
+	git diff --exit-code $@
```
[the remark this answers is not in the document]
Comment (srenatus) [reply components/authz-service/Makefile]: That's not what this catches, though -- it catches if the committed index doesn't match the committed migrations. But yeah, we can improve dev UX 😄

File: components/authz-service/Makefile [hunk 3]
```diff
@@ -14,10 +14,15 @@ ifdef CI
...
 all: lint build test
-
-static: lint
+static: lint ${MIGRATION_READMES}
 unit: build test
 
+.PHONY: ${MIGRATION_READMES}
+${MIGRATION_READMES}:
+	echo '## SQL migrations' > $@
+	for file in $$(dirname $@)/*.up.sql; do f=$$(basename $$file); echo "- [\`$$f\`]($$f): $$(sed '1s/--\/ //;q' "$$file")" >> $@; done
```
[the remark this answers is not in the document]
Comment (srenatus) [reply components/authz-service/Makefile]: The generated README will look odd. I trust developers to care enough about that to not have a check 😃

**Comment by srenatus** [comment]
> B. How is this addressing the root problem, that there are no duplicate sequence numbers?
Whether on my own branch, or someone else doing it on their branch, if I follow the above, one can still commit a duplicate sequence number and the check will pass.
>
> Perhaps I am misunderstanding the suggested workflow?

There's no simple way to hinder two PRs from having overlapping sequence numbers. What we need to stop is **both getting merged**.

So, assuming PR A and PR B have both added migration no. 46; they'll both pass CI and be green. (They'll only get green if they've updated the index README.md.)
However, when (say) PR B is merged, PR A will be detected as *conflicting `master`* by Github. So, the author of PR A needs to react, and they'll figure out that someone else has committed migration 46, and change their number.

I'm open to better ideas, of course 😄

**Comment by srenatus** [comment]
Agreed re: adding some docs about how to add a new migration. Feel free to adopt, since I'm out Monday, if this is becoming a road block.

File: components/authz-service/Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (msorens) [reply components/authz-service/Makefile]: I see the issue as knowing that the readme is touched and needs to be looked at. 
Even from reading your PR, it was not clear to me that that is something I need to be doing from now on.
Having an error emitted here would let folks know.

[the remark this answers is not in the document]
Comment (srenatus) [reply components/authz-service/Makefile]: Yup, taken care of that. Please re-review. 😃

**Pull request merged**